### PR TITLE
feat: Add basic graph domain and more standard library functions for Style

### PR DESCRIPTION
**Pull request by hypotext**
# Description

Related issue/PR: #434 #367 #319 #120

This PR adds a basic graph domain (with pure Penrose layout) and more standard library functions for Style (incl. functions, constraints, objectives, autodiff primitives). 

This reasons are to pilot the user study [e.g. building stdlib, reporting errors], to have a more complicated layout test case, and because it’s just a useful domain to have, since we've been wanting to build it since last summer.

# Implementation strategy and design decisions

For further documentation, see 2/25/21 meeting notes.

# Examples with steps to reproduce them

in `graph-domain`, run `roger watch graph1.sub graph-theory.sty graph-theory.dsl`

image

Added to the working examples page.

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass locally using `yarn test`
- [x] I ran `yarn docs` and there were no errors when generating the HTML site
- [x] My code follows the style guidelines of this project (e.g.: no ESLint warnings)

Also,
- [x] All prior Penrose diagrams (as documented on the wiki page) work

# Open questions

- Improve the graph layout; see TODOs in the Style file
  - Especially address incorporating bbox dimensions (and why it slows down the opt a lot)
- Add more shapes: curved "connector" shapes, rounded boxes, line breaks
- File issues about the misleading errors documented in the Style file

### CODE REVIEW

**Comment by hypotext** [comment]
@wodeni LMK if you want any changes. Getting this in sometime before noon Fri would be great, thanks!

File: packages/core/src/contrib/Constraints.ts [hunk 1]
```diff
@@ -541,64 +628,139 @@ const pointInBox = (p: any, rect: any): boolean => {
...
-  const NUM_SAMPLES = 15;
-  const NUM_SAMPLES2 = constOf(1 + NUM_SAMPLES);
-  // TODO: Check that this covers the whole line, i.e. no off-by-one error
-  const samples = _.range(1 + NUM_SAMPLES).map((i) => {
-    const k = div(constOf(i), NUM_SAMPLES2);
-    return lerp2(line[0], line[1], k);
-  });
-
-  return samples;
+    const NUM_SAMPLES = 15;
```
Comment (wodeni) [inline packages/core/src/contrib/Constraints.ts:655]: Strange spacing. Can you check if you are running Prettier in the pre-commit hook?

**Review by wodeni — APPROVED** [review]

File: packages/core/src/contrib/Constraints.ts (hunk 1 above)
Comment (hypotext) [reply packages/core/src/contrib/Constraints.ts:655]: I thought I was. Will check out later tonight...

Comment (hypotext) [reply packages/core/src/contrib/Constraints.ts:655]: How should I be checking if I'm running Prettier? Maybe we can discuss in the team meeting

Comment (wodeni) [reply packages/core/src/contrib/Constraints.ts:655]: You should get messages like below when the commit hook is running. We can double-check later.
image

Comment (hypotext) [reply packages/core/src/contrib/Constraints.ts:655]: Thanks. Yeah, I'm getting those messages on all my commits (i.e. the hook is running). Not sure why the indentation would be different here given that.

**Pull request merged**